### PR TITLE
Clang's __builtin_memset_inline/__builtin_memcpy_inline introduction.

### PR DESCRIPTION
**Pull request by devnexen**
Unlike their non inline counterparts, no extra calls are generated
at LLVM IR time but work only on compile time sizes so we sort
 of replicate this constraint for other compilers.

### CODE REVIEW

File: rpmalloc/rpmalloc.c [hunk 1]
```diff
@@ -1834,7 +1858,7 @@ _rpmalloc_heap_extract_new_span(heap_t* heap, heap_size_class_t* heap_size_class
 
 static void
 _rpmalloc_heap_initialize(heap_t* heap) {
-	memset(heap, 0, sizeof(heap_t));
+	_rpmalloc_memset_const(heap, 0, sizeof(heap_t));
```
Comment (mjansson) [inline rpmalloc/rpmalloc.c:1861]: Do you have any example of the final generated assembly too see the difference? Is it worth the extra code and platform maintenance?

**Review by mjansson — CHANGES_REQUESTED** [review]

File: rpmalloc/rpmalloc.c [hunk 2]
```diff
@@ -28,6 +28,30 @@
...
+#if !defined(__has_builtin)
+#define __has_builtin(b) 0
+#endif
+
+#if __has_builtin(__builtin_memcpy_inline)
+#define _rpmalloc_memcpy_const(x, y, s) __builtin_memcpy_inline(x, y, s)
+#else
+#define _rpmalloc_memcpy_const(x, y, s)											\
+	do {														\
+		_Static_assert(__builtin_choose_expr(__builtin_constant_p(s), 1, 0), "len must be a constant integer");	\
```
[the remark this answers is not in the document]
Comment (devnexen) [reply rpmalloc/rpmalloc.c:40]: > does gcc have these __builtin:s
no not yet.

File: rpmalloc/rpmalloc.c (hunk 1 above)
Comment (devnexen) [reply rpmalloc/rpmalloc.c:1866]: Here __builtin_memset_inline
`leaq	-9(%rbp), %rax
	movb	$0, 4(%rax)
	movl	$0, (%rax)
`
vs `memset`

`leaq	-9(%rbp), %rdi
	xorl	%esi, %esi
	movl	$5, %edx
	callq	memset@PLT
	leaq	.L.str.3(%rip), %rdi
	movb	$0, %al`

`__builtin_memcpy_inline`

`leaq	-9(%rbp), %rax
	movq	-24(%rbp), %rcx
	movb	4(%rcx), %dl
	movb	%dl, 4(%rax)
	movl	(%rcx), %ecx
	movl	%ecx, (%rax)
	leaq	.L.str.4(%rip), %rdi
	movb	$0, %al`

vs `memcpy`

`leaq	-9(%rbp), %rdi
	movq	-24(%rbp), %rsi
	movq	-32(%rbp), %rdx
	callq	memcpy@PLT
	xorl	%eax, %eax
	addq	$32, %rsp
	popq	%rbp`

So basically the _inline flavor guarantees no call are generated.
